### DISCUSSION TITLE
I would like the CycloneDX components to be sorted by `bom-ref`

### DISCUSSION THREAD

**Discussion by trevor-vaughan** [aquasecurity/trivy#4685] [Ideas]
### Description

When attempting to do a few simple checks in a CI pipeline, I needed to diff a CycloneDX JSON file between a new and old commit.

I discovered that a simple `diff` (ignoring timestamp and global UUID changes) would not work since the `components` are not sorted consistently at each run.

While this is straightforward to work around, it would be great to have sorted output for quick checks.

### Target

Filesystem

### Scanner

None

**Comment by knqyf263** [comment]
Interesting. We do sort components. Could you try the latest version?
https://github.com/aquasecurity/trivy/blob/7d48c5d5d485906e001bdd5103e82767eeebfd7f/pkg/sbom/cyclonedx/core/cyclonedx.go#L206-L208

**Reply by trevor-vaughan** [reply]
I was on 0.41.0 but 0.42.1 has the same result.

```bash
mkdir test
cd test
npm init -y
npm i typescript consola jest
trivy fs --format cyclonedx --output one.json .
sleep 30
trivy fs --format cyclonedx --output two.json .
```

```diff
diff one.json two.json
4c4
<   "serialNumber": "urn:uuid:f655b8f5-3743-4553-810c-56200e3c523c",
---
>   "serialNumber": "urn:uuid:521c28c2-9cd0-4b11-b83a-1f81976a2141",
7c7
<     "timestamp": "2023-06-21T13:07:52+00:00",
---
>     "timestamp": "2023-06-21T13:08:35+00:00",
16c16
<       "bom-ref": "bef63938-6f8d-4fcb-97dc-eb24bcfb4f77",
---
>       "bom-ref": "c7c30f26-c02f-4c20-a13d-318b4ac85c67",
29c29
<       "bom-ref": "65dc0e17-52a7-4349-adb4-26ad2e49e32d",
---
>       "bom-ref": "5df85edf-88f5-4191-9b71-942d0c198434",
5819c5819
<       "ref": "65dc0e17-52a7-4349-adb4-26ad2e49e32d",
---
>       "ref": "5df85edf-88f5-4191-9b71-942d0c198434",
5827c5827
<       "ref": "bef63938-6f8d-4fcb-97dc-eb24bcfb4f77",
---
>       "ref": "c7c30f26-c02f-4c20-a13d-318b4ac85c67",
5829c5829
<         "65dc0e17-52a7-4349-adb4-26ad2e49e32d"
---
>         "5df85edf-88f5-4191-9b71-942d0c198434"
```

I expect the `serialNumber` and `timestamp` fields to differ, but I don't expect the rest to change order at every run.

**Reply by knqyf263** [reply]
It is not a sorting issue. BOM-Ref is UUID, which cannot be consistent.

**Reply by trevor-vaughan** [reply]
@knqyf263 Ah, yes. Apparently reading is hard 🤦.

Appreciate the second set of eyes.

**Reply by trevor-vaughan** [reply]
Actually, wait. Why is `bom-ref` a UUID?

I get that it needs to be unique, but wouldn't it be better to make it a hash of the target if possible so that you could determine at a later date if the read artifact differs?

**Reply by trevor-vaughan** [reply]
Eh, I guess that would be more of a "convenience" feature than anything. Going ahead and closing the discussion as outdated.

**Reply by knqyf263** [reply]
I have no idea why they chose UUID.

> The only requirement for bom-ref is that it is unique within the BOM. Package URL (PURL) is an ideal choice for bom-ref as it will be both unique and readable. If PURL is not an option or not all components represented in the BOM contain a PURL, then UUID is recommended. 

https://cyclonedx.org/use-cases/#dependency-graph

You can propose your idea to CycloneDX.
